exp/internal/controllers/machinepool_controller.go

```diff
@@ -282,12 +287,24 @@ func (r *MachinePoolReconciler) reconcileDeleteNodes(ctx context.Context, cluste
 	return r.deleteRetiredNodes(ctx, clusterClient, machinepool.Status.NodeRefs, machinepool.Spec.ProviderIDList)
 }
 
+// isMachinePoolDeleteTimeoutPassed check the machinePool node delete time out.
+func (r *MachinePoolReconciler) isMachinePoolNodesDeleteTimeoutPassed(machinepool *expv1.MachinePool) bool {
+	if !machinepool.DeletionTimestamp.IsZero() && machinepool.Spec.Template.Spec.NodeDeletionTimeout != nil {
+		return machinepool.DeletionTimestamp.Add(machinepool.Spec.Template.Spec.NodeDeletionTimeout.Duration).After(time.Now())
```

Comment:
if deletiontimestamp+timeout is after Now it means the timeout has not passed and it should return false, no?
Also this func needs to account for 0s as always false.

Can we please have unit tests to cover all these cases?

Reply:
I added more unit test and checking for 0s (Not sure why 0s means continue to delete [forever](https://github.com/kubernetes-sigs/cluster-api/blob/main/api/v1beta1/machine_types.go#L131), not make sense to me should be similar to nodeDrainTimeout and nodeDetachedTimeout immediate delete.)